tests/test_docs_advanced_cast_custom.cpp

```diff
@@ -0,0 +1,86 @@
+// #########################################################################
+// PLEASE UPDATE docs/advanced/cast/custom.rst IF ANY CHANGES ARE MADE HERE.
+// #########################################################################
+
+#include "pybind11_tests.h"
+
+namespace user_space {
+
+struct Point2D {
+    double x;
+    double y;
+};
+
+Point2D negate(const Point2D &point) { return Point2D{-point.x, -point.y}; }
+
+} // namespace user_space
+
+namespace pybind11 {
+namespace detail {
+
+template <>
+struct type_caster<user_space::Point2D> {
+    // This macro inserts a lot of boilerplate code and sets the default type hint to `tuple`
+    PYBIND11_TYPE_CASTER(user_space::Point2D, const_name("tuple"));
+    // `arg_name` and `return_name` may optionally be used to specify type hints separately for
+    // arguments and return values.
+    // The signature of our identity function would then look like:
+    // `identity(Sequence[float]) -> tuple[float, float]`
+    static constexpr auto arg_name = const_name("Sequence[float]");
+    static constexpr auto return_name = const_name("tuple[float, float]");
+
+    // C++ -> Python: convert `Point2D` to `tuple[float, float]`. The second and third arguments
+    // are used to indicate the return value policy and parent object (for
+    // return_value_policy::reference_internal) and are often ignored by custom casters.
+    static handle cast(const user_space::Point2D &number, return_value_policy, handle) {
+        // Convert x and y components to python float
+        auto *x = PyFloat_FromDouble(number.x);
+        auto *y = PyFloat_FromDouble(number.y);
+        // Check if conversion was successful otherwise clean up references and return null
+        if (!x || !y) {
+            Py_XDECREF(x);
+            Py_XDECREF(y);
+            return nullptr;
+        }
+        // Create tuple from x and y
+        auto *t = PyTuple_Pack(2, x, y);
+        // Decrement references (the tuple now owns x an y)
+        Py_DECREF(x);
+        Py_DECREF(y);
+        return t;
+    }
+
+    // Python -> C++: convert a `PyObject` into a `Point2D` and return false upon failure. The
+    // second argument indicates whether implicit conversions should be allowed.
+    bool load(handle src, bool) {
+        // Check if handle is valid Sequence of length 2
```

Comment:
Similarly here ...

Could you please try to use the pybind11 APIs?

Compare with `array_caster::convert_elements()` in pybind11/stl.h:

Roughly (totally untested):

```
        auto l = py::reinterpret_borrow<sequence>(seq);                             
        if (l.size() != 2) {                                                 
            return false;                                                       
        }
        double xy[2];
        ix = 0;                                                                      
        for (auto it : l) {                                                     
            py::make_caster<double> conv;                                                    
            if (!conv.load(it, convert)) {                                      
                return false;                                                   
            }
            xy[ix++] = py::cast_op<double>(conv);                                                                  
        }                                                                       
        value.x = xy[0];
        value.y = xy[1];
        return true;                                                            
```

Reply:
I have changed it to use the pybind11 API.